File: problems/36. Valid Sudoku/main.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        boxes = [set() for _ in range(9)]

        for i in range(9):
            for j in range(9):
                num = board[i][j]
                if num == ".":
                    continue
                if num in rows[i]:
                    return False
                if num in cols[j]:
                    return False
                box_index = (i // 3) * 3 + (j // 3)
                if num in boxes[box_index]:
                    return False

                rows[i].add(num)
                cols[j].add(num)
                boxes[box_index].add(num)

        return True


class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                val = int(board[i][j])
                bit = 1 << val
                box_index = (i // 3) * 3 + (j // 3)

                if rows[i] & bit or cols[j] & bit or boxes[box_index] & bit:
                    return False

                rows[i] |= bit
                cols[j] |= bit
                boxes[box_index] |= bit

        return True

File: problems/36. Valid Sudoku/main.go
func isValidSudoku(board [][]byte) bool {
	rows := make([]int, 9)
	cols := make([]int, 9)
	boxes := make([]int, 9)

	for i := 0; i < 9; i++ {
		for j := 0; j < 9; j++ {
			if board[i][j] == '.' {
				continue
			}
			val := int(board[i][j] - '0')
			bit := 1 << val
			boxIndex := (i/3)*3 + (j / 3)

			if (rows[i]&bit != 0) || (cols[j]&bit != 0) || (boxes[boxIndex]&bit != 0) {
				return false
			}

			rows[i] |= bit
			cols[j] |= bit
			boxes[boxIndex] |= bit
		}
	}

	return true
}
